File: ruby/leetcode019.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end( head, n )
    start = ListNode.new(0)
    start.next, p1, p2 = head, start, head
    n.times { p2 = p2.next }
    while (p2 != nil)
        p1, p2 = p1.next, p2.next
    end
    p1.next = p1.next.next
    start.next
end

File: Archive#00-001-100/019-Remove Nth Node From End of List/leetcode019.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next, p1, p2 = head, dummy, head
        while n > 0:
            p2 = p2.next
            n -= 1
        while p2:
            p1, p2 = p1.next, p2.next
        p1.next = p1.next.next
        return dummy.next
